### ISSUE TITLE
Incorrect and non-deterministic connectedComponents() results when the vertices and edges are not cached

### ISSUE THREAD

**Issue by nchammas** [graphframes/graphframes#159]
I am seeing some surprising and very worrying behavior for `connectedComponents()`. I am running it on a graph with ~300K vertices and ~275K edges. When I run it locally, I get consistent, correct results. When I run it on our cluster, I get inconsistent, incorrect results.

I can't share a repro at the moment, so let me elaborate a bit on how the results are incorrect when I run `connectedComponents()` on our cluster.

When I run `connectedComponents()` on the cluster, about 20-30% of the component IDs returned (so, ~60K-100K) are the same. It's usually a low number from 1 to 10. This is totally incorrect, since I know from running the same code locally and inspecting the results that 200 vertices at most should share the same component ID.

Additionally, when I re-run the same job on our cluster I get different counts every time. So one time it may be ~60K vertices that get assigned a component ID of 0; another time ~80K vertices will be assigned a component ID of 7; and so on. I don't see any such behavior when running locally. When I run this same job locally, I get consistent counts: The most frequently occurring component ID is always shared by exactly 171 vertices.

I don't have any hints about what may be going wrong on the cluster, but occasionally I do see errors like this pop up right around when `connectedComponents()` runs:

```
WARN TransportChannelHandler: Exception in connection from /x.x.x.x:55250
ERROR TransportResponseHandler: Still have 1 requests outstanding when connection from /x.x.x.x:55250 is closed
WARN BlockManagerMaster: Failed to remove broadcast 56 with removeFromMaster = true - Connection reset by peer
```

I don't see them every time though, so they may not be related to the incorrect behavior I'm seeing.

I did figure out a way to work around this problem. If I `persist()` the vertices and edges before calling `connectedComponents()`, I get correct, consistent results like I do when running locally. So **this smells like an optimizer bug in Spark**, loosely related to others like [SPARK-18589](https://issues.apache.org/jira/browse/SPARK-18589) and [SPARK-18492](https://issues.apache.org/jira/browse/SPARK-18492) where adding in a `persist()` at the right place works around the issue.

If someone is interested in tracking this down, I would be happy to work with them to shed more light on what's happening and perhaps boil things down to a reproducible example.

Cluster Environment:
* GraphFrames 0.3.0
* Spark 2.0.2 or Spark 2.1.0
* Scala 2.11
* Java 7
* Python 3.5
* YARN 2.6.0, CDH 5.5.6

Local Environment (same as cluster except for):
* Spark Standalone (no YARN)
* Java 7 or Java 8

**Comment by nchammas** [comment]
Actually, it doesn't look like simply calling `persist()` is enough. Once I scaled up to the full dataset I needed to process (~250M rows), I started seeing the incorrect behavior again.

I was able to again work around it by forcing a computation after the `persist()` with something like `count()`, _before_ calling `connectedComponents()`.

This kind of issue is really discouraging. It's likely not in GraphFrames itself but in the Spark optimizer, and it only shows at scale with a complex plan. It'll probably be very difficult to track this issue down.

**Comment by felixcheung** [comment]
this sounds fairly serious - I think it would be worthwhile to get a non-GraphFrames specific repro

**Comment by nchammas** [comment]
Would the query plans with and without the `persist()` be helpful to start understanding what's going on? They are large and messy though.

**Comment by felixcheung** [comment]
that might be a good starting point..

**Comment by felixcheung** [comment]
that would be great!

**Comment by nchammas** [comment]
Here is the code snipped verbatim from my application where I am seeing this behavior:

```python
    graph = GraphFrame(vertices, edges)
    connected_components = (graph.connectedComponents())
    print("connected_components plan without persist")
    connected_components.explain(extended=True)

    (connected_components
        .groupBy('component')
        .count()
        .orderBy('count', ascending=False)
        .show())

    # IMPORTANT: Required otherwise we get incorrect results on the cluster!
    # See: https://github.com/graphframes/graphframes/issues/159
    vertices = round_trip_to_disk(vertices).persist()
    edges = round_trip_to_disk(edges).persist()
    print('vertices:', vertices.count())
    print('edges:', edges.count())

    graph = GraphFrame(vertices, edges)
    connected_components = (graph.connectedComponents())
    print("connected_components plan with persist")
    connected_components.explain(extended=True)

    (connected_components
        .groupBy('component')
        .count()
        .orderBy('count', ascending=False)
        .show())
```

`round_trip_to_disk()` just writes the DataFrame out to disk and re-reads it. I use it to collapse the lineage graph and avoid optimizer bugs I've been hitting but don't fully understand, including the bug that's reported here.

Here are the plans printed out by the calls to `explain()`, alongside the output I see from counting the most frequently occurring `component` IDs.

Incorrect behavior when I don't persist `vertices` and `edges`:
* Plan: [connected-components-plan-without-persist.txt](https://github.com/graphframes/graphframes/files/709447/connected-components-plan-without-persist.txt)
* Output:

 ```
+---------+--------+
|component|   count|
+---------+--------+
|        0|33619084|
|  1918670|      38|
|    78031|      31|
|  1197110|      30|
|  1721224|      28|
|  3217334|      28|
|   358486|      28|
|  6562825|      27|
|    71512|      27|
|   921120|      26|
|   717668|      26|
|  4069954|      25|
|  1070459|      25|
|  1670368|      25|
|  1867092|      25|
|  1064765|      25|
|   163544|      25|
|  6032386|      25|
|   623762|      25|
|  2061276|      24|
+---------+--------+
 ```

Correct behavior when I do persist `vertices` and `edges`:
* Plan: [connected-components-plan-with-persist.txt](https://github.com/graphframes/graphframes/files/709446/connected-components-plan-with-persist.txt)
* Output:

 ```
+---------+-----+
|component|count|
+---------+-----+
|   127198| 2165|
|   482934| 2131|
|   109805| 1894|
|   296893| 1829|
|   211460| 1769|
|   608799| 1768|
|   905118| 1707|
|   584502| 1706|
|   909496| 1629|
|   133193| 1603|
|   191345| 1572|
|   454205| 1530|
|   340080| 1529|
|   328930| 1464|
|   758840| 1451|
|   357896| 1429|
|   420296| 1415|
|   864493| 1409|
|   716509| 1342|
|   443139| 1327|
+---------+-----+
 ```

As I noted in my original post, I am only seeing this behavior at scale in our cluster environment, and after a complex sequence of operations. So the plans are unfortunately very large and, without anyone willing to spend the time picking them apart, or without specialized tools, it will be very difficult to find out what's going on here.

This has to be one of the most terrifying types of problems to hit in a piece of software--when you silently get incorrect results. I am willing to help track down this problem with guidance from someone familiar with Catalyst, or whatever Spark component may be the culprit here.

**Comment by dingbo8128** [comment]
I also encounter the problem, here my statistics
component  count
0       10880953                                                                
2337725	18
1008415	17
1022156	17
40512	17
240380	16
2584130	15

every time I run the algorithm, I get a different result. It's terrible!

**Comment by nchammas** [comment]
@dingbo8128 - The workaround that worked for me was to round-trip the edges and vertices to disk (e.g. save as Parquet and then immediately load) right before calling `connectedComponents()`. 

The problem appears to be related to the size and complexity of the execution plan. Round-tripping to disk collapses the plan and works around the issue.

@dingbo8128 - Could you share some details about your platform? What versions of Spark, Hadoop, YARN, Scala, Java, Python, etc. are you running?

**Comment by chiangqiqi** [comment]
encounter  with this same problem, bypass using the  `round_trip_to_disk` method.

**Comment by nchammas** [comment]
@dom9988 @dingbo8128 @chiangqiqi - To help track down the source of this issue and perhaps boil it down to a minimal repro, can you share details about your environment where you are seeing this issue?

For example, please share:
* What versions are you using of the following?
  * GraphFrames
  * Spark
  * Hadoop
  * Python / Scala / Java
  * YARN / Mesos (if present)
* What operations are in the execution plan when you see this incorrect behavior?

   For example, I have a suspicion that Python UDFs are a critical component to reproducing this bug. I am curious to know if they are somehow also involved in your cases too.

   Is there any other suspicious operation that, once collapsed using `round_trip_to_disk()` (or some similar technique), resolves the broken `connectedComponents()` behavior for you?

**Comment by chiangqiqi** [comment]
What versions are you using of the following?

- GraphFrames 0.2.0 / 0.2.0

- Spark 2.1.0
- Hadoop 2.7.1
- Scala  2.11.7 

What operations are in the execution plan when you see this incorrect behavior?


```
// orders is retreived from a hive table 
val df = orders.groupBy("id1", "id2").agg(sum("price") as "gmv")

val ids = Seq(0, 1)
  .map(idx ⇒ getColumnsDist(df, idx))
  .reduce(_ union _).toDF("id")

val edges = (df.toDF((Array("src", "dst") ++ df.columns.drop(2)):_*))

val g = GraphFrame(ids, edges)
```
this did cause the error just as @nchammas show , most vertex has component 0, 
after `round_trip_to_disk` before generate the `GraphFrame`, the result is resolved.

**Comment by nchammas** [comment]
That is both interesting and concerning. Seems you don't need such a complex plan to reproduce this issue.

**Comment by rahul1** [comment]
Is anyone planning on fixing this? I have a project that uses graphx that needs to be ported to graphframes, but I am blocked on this issue

**Comment by szhem** [comment]
I'm also interested in this issue. 
Just guessing - is it somehow possible that in the samples above, for example, in this one
```
val ids = Seq(0, 1)
  .map(idx => getColumnsDist(df, idx))
  .reduce(_ union _).toDF("id")
```
there is a standard hadoop's InputFormat involved (which usually reuse key and val instances)? 
If so, I believe, there will be incorrect results in case of any combine-containing operations (like reduceByKey, etc) which follows reading the data.

@dom9988, @dingbo8128, @chiangqiqi, could you please try to run your examples without persisting the vertices and edges (`round_trip_to_disk`) but with calling `repartition(Int)` on vertices and edges dataframes with the number of partitions not equal to the dataframe's one (i.e. obtained using `spark.sql.shuffle.partitions` or `df.rdd.partitions`)?

**Comment by chiangqiqi** [comment]
@szhem amazing
I changed the code to 

```
// orders is retreived from a hive table 
val df = orders.groupBy("id1", "id2").agg(sum("price") as "gmv")

val ids = Seq(0, 1)
  .map(idx ⇒ getColumnsDist(df, idx))
  .reduce(_ union _).toDF("id")

val edges = (df.toDF((Array("src", "dst") ++ df.columns.drop(2)):_*))

val g = GraphFrame(ids.repartition(100), edges.repartition(100))
```

and the result seems to be right.

so can you give a more detailed explaination on what's happening behind this?

**Comment by szhem** [comment]
@chiangqiqi Quite often spark uses standard hadoop [InputFormats](https://hadoop.apache.org/docs/r2.7.3/api/org/apache/hadoop/mapreduce/InputFormat.html) and [RecordReaders](https://hadoop.apache.org/docs/r2.7.3/api/org/apache/hadoop/mapreduce/RecordReader.html) to read the data from hdfs-compatible filesystem.

Standard RecordReaders usually reuse the same instances of key, value for different records of the same InputSplit to prevent GC pressure. Here is [an example of standard KeyValueLineRecordReader](https://github.com/apache/hadoop/blob/f67237cbe7bc48a1b9088e990800b37529f1db2a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/lib/input/KeyValueLineRecordReader.java#L78) that demonstrates such a behavior.

[For RDDs spark notices of such a behavior in the docs](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/SparkContext.scala#L1177)
```
Because Hadoop's RecordReader class re-uses the same Writable object for each record, 
directly caching the returned RDD or directly passing it to an aggregation or shuffle operation 
will create many references to the same object.

If you plan to directly cache, sort, or aggregate Hadoop writable objects, you should first copy
them using a `map` function.
```

[Dataframes may also use standard RecordReaders](https://github.com/apache/spark/blob/b03b4adf6d8f4c6d92575c0947540cb474bf7de1/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/RecordReaderIterator.scala#L56) (that means - reuse instances of the read records).

Re-partitioning seemed to help because it does not use combining (... and event although it seems to help - it's quite ugly workaround) and it still not quite clear whether the issue is connected with the reused records.

To be sure that the issue is connected with the reused records you can try to create a copy of each row instead of re-partitioning, something like this
```
// orders is retreived from a hive table 
val df = orders.groupBy("id1", "id2").agg(sum("price") as "gmv")

val ids = Seq(0, 1)
  .map(idx ⇒ getColumnsDist(df, idx))
  .reduce(_ union _).toDF("id")

val edges = (df.toDF((Array("src", "dst") ++ df.columns.drop(2)):_*))

val g = GraphFrame(ids.map(_.copy()), edges.map(_.copy()))
```
if this sample will work then the issue is definitely in the reused records on some level.

**Comment by nchammas** [comment]
Thanks for the explanation @szhem. Do you know what the equivalent would be in Python?

Is this sufficient to prevent the reuse of record IDs?

```python
>>> spark.range(10).rdd.map(lambda x: x).toDF()
DataFrame[id: bigint]
```

Or do we need something more involved, like this?

```python
>>> import copy
>>> spark.range(10).rdd.map(lambda x: copy.copy(x)).toDF()
DataFrame[id: bigint]
```

**Comment by szhem** [comment]
Hi @nchammas,

I'm not very familiar with pyspark - it uses py4j bridge and it requires additional investigations. 

@chiangqiqi, could you please check whether copying dataframe rows solves the issue?

**Comment by jkbradley** [comment]
@nchammas and others, thanks a lot for reporting these issues and discussing here.  Apologies for the slow response.

A few questions:
* Are your vertex indices Int or Long types?  Or are they something different like String?
  * If they are not integer types, then my first impulse is to blame indexing.  We've hit many bugs with DataFrame and RDD computation of unique IDs, and there are still open bugs in Spark.  E.g., https://issues.apache.org/jira/browse/SPARK-17833  These bugs are often fixable by getting the vertices and edges DataFrames directly from disk, which makes partitioning and row ordering deterministic.
* To confirm, is this using the new implementation of ConnectedComponents in 0.3 and 0.4 (which is based on DataFrames), as opposed to the old implementation which wraps GraphX (but is also available in 0.3 and 0.4)?

**Comment by nchammas** [comment]
@jkbradley:
* My vertex indices are of type `Struct[String, String]`. The PySpark code that builds the vertices DataFrame is basically the following:
  ```python
    vertices = (
        predictions
        .select(
            struct('p1.dataset_name', 'p1.primary_key')
            .alias('id'))
        .union(
            predictions
            .select(
                struct('p2.dataset_name', 'p2.primary_key'))
        )
        .distinct())
  ```
* I am seeing the buggy behavior with ConnectedComponents 0.3 using the new DataFrame-based algorithm, but if I remember correctly I also tried and saw it with the older RDD-based algorithm too.

**Comment by mengxr** [comment]
Based on the behavior, I guess the issue is very likely to be nondeterministic ID assignment. A relevant bug was fix in 2.1.0: https://issues.apache.org/jira/browse/SPARK-14393. In GraphFrames, we convert DataFrames into rdd to get around of this bug. See https://github.com/graphframes/graphframes/blob/master/src/main/spark-2.x/org/apache/spark/sql/SQLHelpers.scala#L33. However, if the source DataFrame is not deterministic, e.g., after shuffle, the ID assignment won't be stable.

@jkbradley I think we should consider a more expensive fix:
* hash partitioning
* sort within each partition + dedup
* assign IDs after

This requires shuffle and sort. But if we use many partitions, sorting within each should be cheap. It is definitely slower, but better than inaccurate.

**Comment by mengxr** [comment]
@nchammas I will create a PR to fix it. Could you help test after? Thanks!

**Comment by nchammas** [comment]
@mengxr - I would be happy to help test a fix. FWIW, note that I see the buggy behavior with Spark 2.1.0 too. I haven't tested with 2.1.1 yet.

**Comment by mengxr** [comment]
That's expected. We fixed the bug in 2.1.0, but we didn't switch to use it in graphframes. Because we need to support multiple spark versions here.

**Comment by nchammas** [comment]
Ah, understood. Hence your note about the more expensive fix. 👍

**Comment by mengxr** [comment]
@nchammas I sent a patch at #189.

**Comment by mengxr** [comment]
Hi all, #189 should fix this issue with a more expensive but reliable way to assign long IDs. We will make a bug fix release after it gets merged.

**Comment by mengxr** [comment]
FYI, we merged the fix into master (#195).

**Comment by nchammas** [comment]
I suppose we can close this issue as resolved now that [GraphFrames 0.5.0 has been released](https://github.com/graphframes/graphframes/releases/tag/release-0.5.0).

**Comment by mengxr** [comment]
Thanks everyone for reporting the issue and sorry for the late follow-up!

**Comment by happyelma7** [comment]
@mengxr  in my job, using grphx connectedcomponets() , the result is two vertexs of an edge mapping different graph id in spark 2.1.2,can you help me to solve this question,it happened about one year，please!

**Comment by Fungie** [comment]
I believe I'm still seeing this bug. I'm using graphframes 0.6 with spark 2.3 and scala 2.11. If I don't cache the vertex and edge dataframes before running connectedcomponents the output columns components is inconsistent and duplicates component number.

Example using the public febrl dataset.

Output when not cached

![image](https://user-images.githubusercontent.com/16100465/50765136-e472af80-126c-11e9-893a-083a022ff850.png)

Output when cached

![image](https://user-images.githubusercontent.com/16100465/50765086-b7260180-126c-11e9-9363-364bfc1c5999.png)

**Comment by mengxr** [comment]
@happyelma7 @Fungie Did you use setAlgorithm("graphframes")?

**Comment by Fungie** [comment]
@mengxr @happyelma7 I used `connected = graph.connectedComponents(algorithm='graphframes')`

Is that the same as calling `setAlgorithm("graphframes")`?

**Comment by ronflomin** [comment]
Hi, does anyone have a solution? I tried everything, caching the vertices and edges before creating the graph, writing them to disk and than reading it, nothing works. I uses GraphFrames 0.7

**Comment by ronflomin** [comment]
> FYI, we merged the fix into master (#195).

Hi, are you sure it is fixed?I tried everything, caching the vertices and edges before creating the graph, writing them to disk and than reading it, nothing works. I use GraphFrames 0.7, and most of the vertices are assigned to component 0

**Comment by ronflomin** [comment]
> FYI, we merged the fix into master (#195).

@felixcheung

Hi, are you sure it is fixed?I tried everything, caching the vertices and edges before creating the graph, writing them to disk and than reading it, nothing works. I use GraphFrames 0.7, and most of the vertices are assigned to component 0

**Comment by ronflomin** [comment]
> > FYI, we merged the fix into master (#195).
> 
> @felixcheung
> @mengxr
> 
> Hi, are you sure it is fixed?I tried everything, caching the vertices and edges before creating the graph, writing them to disk and than reading it, nothing works. I use GraphFrames 0.7, and most of the vertices are assigned to component 0

Is anyone still facing this issue?

**Comment by nchammas** [comment]
@ronflomin -

1. Can you boil your problem down to a minimal reproduction of the issue?
2. Are you sure that your results are incorrect? How do you know?
3. What version of Spark are you using?

**Comment by niliafsari** [comment]
Hi, I am also facing this issue too. I'm using Graphframes 0.6 and Spark 2.2. Most of my vertices are assigned to component 0 and I know this is wrong, I checked it with Networkx.  @felixcheung
@mengxr @nchammas @ronflomin

**Comment by niliafsari** [comment]
I have ~9 million edges and ~8 million nodes with many disconnected components

+-------+---------+
|     id|component|
+-------+---------+
|1000839|        0|
|1000963|        0|
|1006113|        0|
|1013839|        0|
|1017547|        0|
| 101825|        0|
|1018582|      480|
|1019301|        0|
|1020848|        0|
|1024379|        0|
|1026210|        0|
|1026815|        0|
|1028527|        0|
|1030466|      792|
|1032351|      852|
|1033095|        0|
|  10341|        0|
|1036553|        0|
|1038462|        0|
|1044108|        0|
+-------+---------+

**Comment by deepaksing05** [comment]
I was able to fix this issue. even I was getting millions of records having Component 0 id. Initially I converted all my vertex ids in alphanumeric format by hashing each value with Sha2 function. Once done I found that the component 0  had less than 10 records and I thought I had fixed it. Later I found that the Vertices connected with Component 0 had been assigned to a different Component id. Therefore the problem was not fixed.

Then after thoroughly analyzing the joined Vertices with Component 0, I found that there was a rogue id 00000-00000-00000-00000 that was joining unrelated subgraphs once Connected Component algorithm was executed. Once I filtered the vertex with  00000-00000-00000-00000 value, my Connected Graph gave me perfect result.
Therefore I would suggest people to look for rogue Vertex values that are common among multiple graphs which leads to millions of vertices joining with Component 0 id. Hope this helps! I had used graphframes-0.7.0-spark2.3-s_2.11.jar Graphframe package.

**Comment by girbeabogdan** [comment]
Hello folks! Dropping a qq here  @mengxr , @jkbradley , @thunterdb 
We're using graphframes 0.6 & spark 2.4.4 and running ConnectedComponents for 335 **billion** edges thus generating 361 **billion** connected components identifiers.

We noticed the ConnectedComponents algorithm is NOT doing the CC assignment for a few vertexes connected by an edge. The problem looks transitive since it gets resolved by following job runs pointing to some dataloss at some point. 

Any ideas on why this would be happening?

**Comment by nchammas** [comment]
@girbeabogdan - That sounds like a separate question unrelated to the subject of this issue. Please open a separate issue.

I'm going to close this issue again due its age and inactivity. The latest release of GraphFrames is [0.8.1](https://github.com/graphframes/graphframes/releases/tag/v0.8.1).

@niliafsari / @ronflomin / others: If you are still seeing the issue I [demonstrated here](https://github.com/graphframes/graphframes/issues/159#issuecomment-272973373) with this latest release of GraphFrames, and you're sure you don't just have some bad data as [suggested here](https://github.com/graphframes/graphframes/issues/159#issuecomment-631026331), then please post an update.

**Comment by liangshi036** [comment]
for connectedComponents, the result seems right after repartion

`component_id    cnt                                                             
6004    8126
8589935457      5093
2640    4561
11783   4015
1458    3974
20983   3696
34660   3514
19124   3482
6700    3417
753     3080
.....`

BUT I still get this wrong result in stronglyConnectedComponents.

even I tried .repartion  or  .map(_copy())  method

`component       cnt                                                             
79      10149
24652   11
68719501839     10
17179905791     7
17179889315     7
8831    7
8589993074      7
42949681311     6
3509    6
94489294169     6
60129553937     6
25769810254     6`

**Comment by i100van** [comment]
Hello good evening,

I am suffering from the same problem, after generating the connected components, a component is generated that attaches many nodes that are unconnected. I have tried the options mentioned in the thread (collapse the spark graph using the round trip, make a previous count, use a repartition) but nothing works, I keep receiving as output a component that agglutinates millions of nodes, and whose number is usually 0 but oscillates.

I have no more ideas to resort to, I don't know if anyone has any idea that could help, it would be of great help.

Thank you very much for your help.

**Comment by deepaksing05** [comment]
> Hello good evening,
> 
> I am suffering from the same problem, after generating the connected components, a component is generated that attaches many nodes that are unconnected. I have tried the options mentioned in the thread (collapse the spark graph using the round trip, make a previous count, use a repartition) but nothing works, I keep receiving as output a component that agglutinates millions of nodes, and whose number is usually 0 but oscillates.
> 
> I have no more ideas to resort to, I don't know if anyone has any idea that could help, it would be of great help.
> 
> Thank you very much for your help.

I would recommend check the nodes that are connected although they should not be. It might be that they have a junk value  which is connecting the edges. For example, you would have already filtered nodes with a NULL value so that dissimilar edges with NULL nodes dont connect. Check for similar values, like spaces or zeros or 1s, as nodes having these common values will be connected. Best of luck!

**Comment by BenMizrahiPlarium** [comment]
+1

**Comment by nixuri** [comment]
+1

**Comment by zhouhanc** [comment]
+1 still had this problem using GraphX on Spark 3.3.2, Scala 2.12.18

**Comment by ZackUhlenhuth** [comment]
Still facing this issue with GraphFrames 0.8.2. I need to round trip to s3 in order to get the correct connectedComponents behavior (it isn't enough to .cache or .persist)

**Comment by boliu68** [comment]
Still has this problem with GraphFrames 0.8.0

**Comment by minerjaime** [comment]
My findings:

We have been experiencing issues with inconsistent results from GraphFrames' connected components implementation:
When connected components runs, the # of nodes within the input graph should equal the # of nodes specified within the output.

We (as well as other GraphFrames users) experience inconsistencies, where these two numbers are unequal.

I believe that the cause of this issue is an incorrect interpretation of the original Google Research paper (https://dl.acm.org/doi/pdf/10.1145/2670979.2670997) and GraphFrames' inability to utilize provided node IDs.

The interpretation errors arise here within the GraphFrames source: https://github.com/graphframes/graphframes/blob/0b4df70038a4f0ff3b4544223089084fc8742da7/src/main/scala/org/graphframes/lib/ConnectedComponents.scala#L364

That currSum​ value is used to terminate the loop.  It is a mis-coded interpretation of the value within Lemma 8​ of the Google Research paper.  The only time that the GraphFrames metric and Lemma 8​ are guaranteed to agree is on the final​ iteration (once convergence has already been achieved).

They can disagree within previous iterations and, moreover, it is likely that the disagreement between the implementations is a cause for the early termination of the algorithm and the production of inconsistent results.  This disagreement would likely be avoided if the GraphFrames implementation accepted provided node IDs, rather than always defaulting to computing its own via Spark's monotonically_increasing_id()​.


The fix is straightforward:
- Rather than use ee​ to compute the check sum, minNbrs1​ should be used.
- The value of the check sum should be as follows (in pyspark​ – I'm not a scala​ person):
`currSum = minNbrs1.select(sum(col(MIN_NBR))).collect()[0][0]`

Also, an outstanding issue with GraphFrames is its messiness with respect to caching and persistence.  I have confirmed this within pyspark​: when you repeatedly cache a DataFrame​ (for example, ee​ is persisted several times on lines: 310, 323, 337), you lose the ability to unpersist​ it's multiple copies.  

This leads to orphaned data with the Spark cache that we cannot remove without clearing the cache in its entirety, which prevents us from using the cache for other applications on the same cluster.

**Comment by ZackUhlenhuth** [comment]
In my case, the problem was actually that we were ordering ids in a nondeterministic way before running connected components (e.g. by ordering a dataframe on a field where many values are equivalent). I believe this causes an issue with connected components giving inconsistent results bc the nodes do not deterministically hash to certain partitions. So it wasn't really an issue with GraphFrames in our case, but it was confusing to debug!

**Comment by olbapjose** [comment]
Has this been totally solved? I am still having the same issue with version 0.8.4 on Databricks. I get a really huge component with id 0, even though I am creating the GraphFrame object as follows (in a Databricks notebook).

```
spark.conf.set("spark.sql.adaptive.enabled", "false")
g = GraphFrame(v=spark.read.table("catalog.database.nodes"), e=spark.read.table("catalog.database.edges")
```

Apparently the code I can see in the master branch of the graphframes repository already has the fix suggested previously in this thread. Furthermore, Databricks incorporates their own optimized GraphFrames package (using DBR 15.4 LTS ML with graphframes 0.8.4-db1-spark3.5) but I get the same result (although it runs 4~5 times faster than the open source package). Both the open source and the Databricks package exhibit the same behaviour.

**Comment by minerjaime** [comment]
@olbapjose Having an extremely large component isn't necessarily an indicator of the issue.  Does GraphFrames consistently return that same component for successive runs?

**Comment by olbapjose** [comment]
Turns out I had an error in my code. The massive component with id 0 was there, but it was deterministic. After solving my error, there's no massive component anymore, thanks.
